File: javascript/working/euler0684.js
import BigNumber from 'bignumber.js';

const MODULO = 1000000007;
// const MAX_SMALLEST_SUM_DIGITS = 999999999;
// const MAX_SMALLEST_SUM = 81;

const SMALLEST_SUM_MAP = new Map();
// const SUM_OF_SMALLEST_SUMS_MAP = new Map();
const FIB_MAP = new Map();

function findSmallestSum(n) {
  if (!SMALLEST_SUM_MAP.get(n)) {
    const num = (n % 9).toString() + new Array(Math.floor(n / 9) + 1).join('9');
    SMALLEST_SUM_MAP.set(n, new BigNumber(num));
  }

  return SMALLEST_SUM_MAP.get(n);
}

// function findSumOfSmallestSums(n) {
//   if (!SUM_OF_SMALLEST_SUMS_MAP.get(n)) {
//     let start = SUM_OF_SMALLEST_SUMS_MAP.size || 1;
//     do {
//       SUM_OF_SMALLEST_SUMS_MAP.set(
//         start,
//         findSmallestSum(start)
//           .plus(SUM_OF_SMALLEST_SUMS_MAP.get(start - 1) || 0)
//           .modulo(MODULO)
//       );
//       start++;
//     } while (start <= n);
//   }
//   return SUM_OF_SMALLEST_SUMS_MAP.get(n);
// }

function findFib(n) {
  if (n <= 0) {
    FIB_MAP.set(0, new BigNumber(0));
    return FIB_MAP.get(0);
  }
  if (n === 1) {
    FIB_MAP.set(1, new BigNumber(1));
    return FIB_MAP.get(1);
  }
  const fib1 = FIB_MAP.get(n - 1) || findFib(n - 1);
  const fib2 = FIB_MAP.get(n - 2) || findFib(n - 2);
  const fib = fib1.plus(fib2);
  FIB_MAP.set(n, fib);
  return fib;
}

// function findFibSum(n) {
//   const fib = findFib(n);
//   return findSumOfSmallestSums(fib);
// }

export function solve684() {
  let n = 20;
  let sum = new BigNumber(0);
  let start = new BigNumber(1);
  let end = new BigNumber(0);

  for (let i = 2; i <= n; i++) {
    let currentSum = new BigNumber(0);
    start = end.plus(1);
    end = findFib(i);
    for (let k = start; k <= end; k++) {
      currentSum = currentSum.plus(findSmallestSum(k).times(n + 1 - i)).modulo(MODULO);
    }

    sum = sum.plus(currentSum).modulo(MODULO);
  }

  return sum.toString();
}

// solve684();

File: python/working/euler0684.py
import math

MODULO = 1000000007
SMALLEST_SUM_MAP = {}
FIB_MAP = {}

def findSmallestSum (n):
  if n not in SMALLEST_SUM_MAP:
    x = math.floor(n / 9)
    num = ((n % 9) * (10 ** x)) + ((10 ** x) - 1)
    SMALLEST_SUM_MAP[n] = num
  
  return SMALLEST_SUM_MAP[n]

def findFib(n):
  if n <= 0:
    FIB_MAP[0] = 0
    fib = 0
  elif n == 1:
    FIB_MAP[1] = 1
    fib = 1
  else:
    if n - 1 in FIB_MAP:
      fib1 = FIB_MAP[n - 1]
    else:
      fib1 = findFib(n - 1)
    if n - 2 in FIB_MAP:
      fib2 = FIB_MAP[n - 2]
    else:
      fib2 = findFib(n - 2)
    fib = fib1 + fib2
    FIB_MAP[n] = fib
    
  return fib

def main(n):
  sum = 0
  start = 0
  end = 0

  for i in range(2, n + 1):
    currentSum = 0
    start = end + 1
    end = findFib(i)
    for k in range(start, end + 1):
      currentSum += (findSmallestSum(k) * (n + 1 - i))

    sum += currentSum

  return sum % MODULO

# for n in range(1, 15):
#   print(n, ':', main(n))

print(main(24))